src/tallies/filter_mesh.cpp

```diff
@@ -0,0 +1,132 @@
+#include "openmc/tallies/filter_mesh.h"
+
+#include <sstream>
+
+#include "openmc/capi.h"
+#include "openmc/constants.h"
+#include "openmc/error.h"
+#include "openmc/mesh.h"
+#include "openmc/xml_interface.h"
+
+namespace openmc {
+
+void
+MeshFilter::from_xml(pugi::xml_node node)
+{
+  auto bins_ = get_node_array<int32_t>(node, "bins");
+  if (bins_.size() != 1) {
+    fatal_error("Only one mesh can be specified per " + type()
+                + " mesh filter.");
+  }
+
+  auto id = bins_[0];
+  auto search = model::mesh_map.find(id);
+  if (search != model::mesh_map.end()) {
+    set_mesh(search->second);
+  } else{
+    std::stringstream err_msg;
+    err_msg << "Could not find mesh " << id << " specified on tally filter.";
+    fatal_error(err_msg);
+  }
+}
+
+void
+MeshFilter::get_all_bins(const Particle* p, int estimator, FilterMatch& match)
+const
+{
+  if (estimator != ESTIMATOR_TRACKLENGTH) {
+    auto bin = model::meshes[mesh_]->get_bin(p->r());
+    if (bin >= 0) {
+      match.bins_.push_back(bin);
+      match.weights_.push_back(1.0);
+    }
+  } else {
+    model::meshes[mesh_]->bins_crossed(p, match.bins_, match.weights_);
+  }
+}
+
+void
+MeshFilter::to_statepoint(hid_t filter_group) const
+{
+  Filter::to_statepoint(filter_group);
+  write_dataset(filter_group, "bins", model::meshes[mesh_]->id_);
+}
+
+std::string
+MeshFilter::text_label(int bin) const
+{
+  auto& mesh = *model::meshes[mesh_];
```

Comment:
If `mesh_` is not set appropriately, this dereference operation will cause a crash.